embeddings/src/lib.rs

```diff
@@ -38,97 +37,203 @@ fn handle_embeddings(req: Request) -> Result<Response> {
     type Error = anyhow::Error;
 
     fn try_from(row: sqlite::Row<'a>) -> std::result::Result<Self, Self::Error> {
-        info!("DB Row: {:?}", row.get::<&str>("embedding"));
-        info!("DB Row: {:?}", row.get::<u32>("embedding"));
-        let id = row.get::<u32>("id").unwrap();
+        // TODO: Fix unwraps()
+        let id = Some(row.get::<u32>("id").unwrap());
         let reference = row
             .get::<&str>("reference")
             .context("reference column is empty")?;
         let text = row.get::<&str>("text").context("text column is empty")?;
-        let embedding = row.get::<&str>("embedding").unwrap_or_default();
-        info!("Embedding from DB {:?}", embedding);
-        //let embedding: Option<Vec<f32>> = embedding.unwrap()
-        //let embedding: Vec<f32> = bincode::deserialize_from(Cursor::new(embedding))?;
+        let embedding: Vec<f32> = match row.get::<&ValueResult>("embedding").unwrap() {
+            ValueResult::Blob(b) => serde_json::from_value(serde_json::from_slice(b.as_slice()).unwrap())?,
+            _ => todo!(),
+        };
         Ok(Self {
             id,
-            reference: reference.to_owned(),
+            reference: Some(reference.to_owned()),
             text: text.to_owned(),
-            embedding: None,
+            embedding: Some(embedding),
         })
     }
 }
+
+fn get_similar(query: Query) -> Result<SimilarityResultSet> {
+    let text: Vec<&str> = vec![query.text.as_str()];
+    let embedding = generate_embeddings(AllMiniLmL6V2, &text[..]);
+    let sql_query = "SELECT * FROM embeddings";
+    let conn = Connection::open_default()?;
+    let embedding_records: Vec<Embedding> = conn
+        .execute(sql_query, &[])?
+        .rows()
+        .map(|row| -> anyhow::Result<Embedding> { row.try_into() })
+        .collect::<anyhow::Result<Vec<Embedding>>>()?;
+
+    let mut result_set = SimilarityResultSet {
+        text: query.text.to_string(),
+        results: Vec::new(),
+    };
+
+    for e in embedding_records.iter() {
+        let similarity = cosine_similarity(
+            &e.embedding.clone().unwrap(),
+            embedding.clone().unwrap().embeddings.get(0).unwrap(),
+        );
+        let mut result = SimilarityResult {
+            similarity,
+            embedding: e.clone(),
+        };
+        result.embedding.embedding = None;
+        result_set.results.push(result);
+    }
+
+    result_set
+        .results
+        .sort_by(|a, b| b.similarity.partial_cmp(&a.similarity).unwrap());
+
+    Ok(result_set)
+}
+
+fn cosine_similarity(vec1: &[f32], vec2: &[f32]) -> f32 {
+    let dot_product = vec1
+        .iter()
+        .zip(vec2.iter())
+        .map(|(x, y)| x * y)
+        .sum::<f32>();
+    let norm1 = vec1.iter().map(|x| x * x).sum::<f32>().sqrt();
+    let norm2 = vec2.iter().map(|y| y * y).sum::<f32>().sqrt();
+    dot_product / (norm1 * norm2)
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+struct Embedding {
+    id: Option<u32>,
```

Comment:
I'm surprised this works without the following 

```suggestion
   #[default]
    id: Option<u32>,
```

It seems like maybe you're trying to use this struct in many different places hence the optional fields (which are never actually optional - they must be set in some uses and must not be set in others?)

Reply:
Yeah, I'm struggling with that part. I guess it doesn't make sense to save a few Structs?

Reply:
I don't think it makes sense to try to use one struct for many different things.

Reply:
Yeah, I can see that. And that brings us back to the whole naming thing :-) But I agree the more I struggle reading this code myself.